File: tests/panel_measles/R_comparison/mixed/logLik_comparison/panel_measles.R
# --- SLURM CONFIG ---
# sbatch_args:
#   job-name: "panel measles loglik comparison (R panelPomp)"
#   partition: standard
#   nodes: 1
#   ntasks-per-node: 36
#   cpus-per-task: 1
#   mem-per-cpu: 2GB
#   output: "results/logs/slurm-%j.out"
# run_levels:
#   1:
#     sbatch_args: { time: "00:02:00" }
#   2:
#     sbatch_args: { time: "00:30:00" }
#   3:
#     sbatch_args: { time: "02:00:00" }
#   4:
#     sbatch_args: { time: "08:00:00" }
# setup: |
#   module load R/4.4.0
# command: |
#   R CMD BATCH --no-restore --no-save panel_measles.R results/logs/panel_measles.Rout
# --- END SLURM CONFIG ---

stopifnot(getRversion() >= "4.1")
library(tidyverse)
library(pomp)
library(panelPomp)
library(doParallel)
library(foreach)
library(doRNG)

RUN_LEVEL <- as.numeric(Sys.getenv("RUN_LEVEL", unset = 1))
NP_EVAL <- switch(RUN_LEVEL, 2, 1000, 5000, 5000)
NREPS_EVAL <- switch(RUN_LEVEL, 2, 300, 300, 3600)

source("../panel_measles_shared.R")


# Load MLE parameters and build panel parameters
mle_params <- read.csv("../../../../measles/R_comparison/data/AK_mles.csv", stringsAsFactors = FALSE)
mles <- mle_params %>% filter(town %in% units)

shared_names <- c("R0", "sigma", "gamma", "sigmaSE", "cohort", "amplitude")
specific_names <- c("iota", "rho", "psi", "S_0", "E_0", "I_0", "R_0")

shared_params <- sapply(shared_names, function(p) mean(mles[[p]]))

specific_params_matrix <- matrix(
  NA,
  nrow = length(specific_names),
  ncol = length(units),
  dimnames = list(specific_names, units)
)
for (u in units) {
  unit_mle <- mles %>% filter(town == u)
  for (p in specific_names) {
    specific_params_matrix[p, u] <- unit_mle[[p]]
  }
}

# Assign to panelPomp object
shared(panel_obj) <- shared_params
specific(panel_obj) <- specific_params_matrix

# Setup Parallel
cores <- as.numeric(Sys.getenv("SLURM_NTASKS_PER_NODE", unset = NA))
if (is.na(cores)) {
  cores <- detectCores()
  if (is.na(cores)) cores <- 1
}
registerDoParallel(cores)
registerDoRNG(594709947L)

# Run Panel Pfilter
print(sprintf("Running %d panel pfilters with Np = %d...", NREPS_EVAL, NP_EVAL))

t_pf_start <- Sys.time()
all_logliks <- foreach(
  i = 1:NREPS_EVAL,
  .packages = c("pomp", "panelPomp"),
  .combine = rbind,
  .options.multicore = list(set.seed = TRUE)
) %dopar% {
  pf_res <- pfilter(panel_obj, Np = NP_EVAL)
  ull <- unitLogLik(pf_res)
  data.frame(
    replicate = i,
    unit = names(ull),
    logLik = as.numeric(ull)
  )
}
t_pf_end <- Sys.time()
pf_time_total <- as.numeric(t_pf_end - t_pf_start, units = "secs")

dir.create("results", recursive = TRUE, showWarnings = FALSE)
saveRDS(all_logliks, "results/pfilter_logliks_f64.rds")
print("Saved results/pfilter_logliks_f64.rds")

# Save timing to CSV
timings_df <- data.frame(
  phase = c("pfilter"),
  time_seconds = c(pf_time_total)
)
write.csv(timings_df, "results/r_pomp_time.csv", row.names = FALSE)
print(sprintf("Pfilter Total Time: %.2f s", pf_time_total))

File: tests/panel_measles/R_comparison/mixed/logLik_comparison/panel_measles.py
# --- SLURM CONFIG ---
# importance: medium
# sbatch_args:
#   job-name: "panel measles loglik comparison (pypomp)"
#   partition: gpu-rtx6000
#   gpus: "rtx_pro_6000_blackwell:1"
#   cpus-per-gpu: 1
#   mem: 6GB
#   output: "results/logs/slurm-%j.out"
# run_levels:
#   1:
#     sbatch_args: { time: "00:04:00" }
#   2:
#     sbatch_args: { time: "00:15:00" }
#   3:
#     sbatch_args: { time: "00:15:00" }
#   4:
#     sbatch_args: { time: "08:00:00" }
# --- END SLURM CONFIG ---

import os
import pickle
from typing import Any
import jax
import numpy as np
import pandas as pd
import pypomp as pp

print(jax.devices())

MAIN_SEED = 594709947
key = jax.random.key(MAIN_SEED)
np.random.seed(MAIN_SEED)

RUN_LEVEL = int(os.environ.get("RUN_LEVEL", "1"))
NP_EVAL = (2, 1000, 5000, 5000)[RUN_LEVEL - 1]
NREPS_EVAL = (2, 300, 300, 3600)[RUN_LEVEL - 1]

print(f"Running at level {RUN_LEVEL}")
print(f"NP_EVAL: {NP_EVAL}, NREPS_EVAL: {NREPS_EVAL}")

# Units to process
CHOSEN_UNITS = ["London", "Halesworth", "Hastings", "Cardiff"]
SHARED_PARAMS = ["R0", "sigma", "gamma", "sigmaSE", "cohort", "amplitude"]
SPECIFIC_PARAMS = ["iota", "rho", "psi", "S_0", "E_0", "I_0", "R_0"]

# Load MLE parameters from CSV
mle_params = pd.read_csv("../../../../measles/R_comparison/data/AK_mles.csv")

# Extract MLEs for the chosen units
mles = mle_params.set_index("town").loc[CHOSEN_UNITS]

# Average the shared parameters across the 4 units
shared_dict = {}
for p in SHARED_PARAMS:
    shared_dict[p] = float(mles[p].mean())
shared_df = pd.DataFrame.from_dict(shared_dict, orient="index").rename(
    columns={0: "shared"}
)

# Specific parameters matrix
specific_df = mles[SPECIFIC_PARAMS].T

# Construct PanelParameters
theta_dict = {"shared": shared_df, "unit_specific": specific_df}
theta = pp.PanelParameters(theta=theta_dict)

# Construct panelPomp
dummy_initial_params_list = pp.Pomp.sample_params(
    {
        p: (
            theta_dict["shared"].loc[p, "shared"],
            theta_dict["shared"].loc[p, "shared"] + 1e-5,
        )
        for p in SHARED_PARAMS
    }
    | {
        p: (
            float(theta_dict["unit_specific"].loc[p].iloc[0]),
            float(theta_dict["unit_specific"].loc[p].iloc[0]) + 1e-5,
        )
        for p in SPECIFIC_PARAMS
    },
    1,
    key=key,
)

pomp_dict = {
    unit: pp.models.UKMeasles.Pomp(
        unit=unit,
        theta=dummy_initial_params_list,
        model="001b",
        clean=False,
    )
    for unit in CHOSEN_UNITS
}

# Override covariates with R-computed ones to match R exactly
import sys

sys.path.append("..")
from align_covariates import align_covariates

align_covariates(pomp_dict)


panel_measles_obj = pp.PanelPomp(
    Pomp_dict=pomp_dict,
    theta=theta,
)


# Run Panel Pfilter
print(f"Running {NREPS_EVAL} panel pfilters with J={NP_EVAL}...")
key, subkey = jax.random.split(key)
import time

t_start = time.perf_counter()
panel_measles_obj.pfilter(J=NP_EVAL, reps=NREPS_EVAL, key=subkey)
t_pfilter = time.perf_counter() - t_start
print(f"Pfilter completed in {t_pfilter:.2f} seconds.")

# Extract logLiks
all_logliks = []
if len(panel_measles_obj.results_history) > 0:
    last_res: Any = panel_measles_obj.results_history[-1]
    logliks_da = last_res.logLiks  # coordinates: theta_idx, unit, rep

    # We only have one parameter set (theta_idx = 0)
    logliks_df = logliks_da.isel(theta_idx=0).to_dataframe(name="logLik").reset_index()

    # logliks_df has columns: unit, rep, logLik
    logliks_df["replicate"] = logliks_df["rep"] + 1

    os.makedirs("results", exist_ok=True)
    output_file = "results/pfilter_logliks_f32.pkl"
    with open(output_file, "wb") as f:
        pickle.dump(logliks_df[["unit", "replicate", "logLik"]], f)

    print(f"Saved results to {output_file}")

    # Save timing to CSV
    timing_df = pd.DataFrame({"phase": ["pfilter"], "time_seconds": [t_pfilter]})
    timing_df.to_csv("results/pypomp_time.csv", index=False)
    print("Saved timing to results/pypomp_time.csv")
else:
    print("Warning: No results in results_history")
